src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,113 @@ public ProjectGraph(string entryProjectFile)
+            Dictionary<ConfigurationMetadata, NodeState> nodeState,
+            ProjectCollection projectCollection,
+            PropertyDictionary<ProjectPropertyInstance> globalProperties)
         {
-            var projectsToEvaluate = new Queue<KeyValuePair<ConfigurationMetadata, ProjectGraphNode>>();
-            // entry project node has no parent
-            projectsToEvaluate.Enqueue(new KeyValuePair<ConfigurationMetadata, ProjectGraphNode>(new ConfigurationMetadata(entryProjectFile, globalProperties), null));
-            while (projectsToEvaluate.Count != 0)
+            nodeState[projectToEvaluate] = NodeState.InProcess;
+            ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+            IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.Project.GetItems(ProjectReferenceItemName);
+            foreach (var projectReferenceToParse in projectReferenceItems)
             {
-                KeyValuePair<ConfigurationMetadata, ProjectGraphNode> projectToEvaluateAndParentNode = projectsToEvaluate.Dequeue();
-                ConfigurationMetadata projectToEvaluate = projectToEvaluateAndParentNode.Key;
-                ProjectGraphNode parentNode = projectToEvaluateAndParentNode.Value;
+                if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                {
+                    throw new InvalidOperationException(string.Format(
+                        CultureInfo.InvariantCulture,
+                        ResourceUtilities.GetResourceString(
+                            "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                        projectReferenceToParse.EvaluatedInclude,
+                        parsedProject.Project.FullPath));
+                }
 
-                if (!_allParsedProjects.TryGetValue(projectToEvaluate, out ProjectGraphNode parsedProject))
+                string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+
+                PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, globalProperties);
+                var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                if (nodeState.TryGetValue(projectReferenceConfigurationMetadata, out NodeState projectReferenceNodeState))
                 {
-                    parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
-                    IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.Project.GetItems(ProjectReferenceItemName);
-                    foreach (var projectReferenceToParse in projectReferenceItems)
+                    // a project reference can be in "Processed" state. If it is "InProcess" state, it is an ancestor and there is a circular dependency
```

Comment:
Can you clarify this comment? Maybe something like

> Because this is a depth-first search, we should only encounter new nodes or nodes whose subgraph has been completely processed. If we encounter a node that is currently being processed, it must be one of our ancestors in a circular dependency.